### PR TITLE
[SourceKit/InterfaceGen] Also print the cross-import overlays of a module in its interface.

### PR DESCRIPTION
**Pull request by nathawes**
When printing the generated interface of a module, also print the decls from any underscored cross-import overlays it is the direct, or indirect underlying module of. Declarations are grouped by overlay, with a `MARK:` comment introducing each overlay, and a regular comment above each decl listing the required bystander modules that must be imported for the decl to be
available.

In addition, in each overlay:
- import declarations of the underlying modules are filtered out, since those modules are either other underscored cross-import overlays, or the target module the overlays are presented as being a part of.
- import declarations of modules that are also imported in the target module are filtered out, also because the overlay is presented as a conditional part of the target module.

@brentdax does that handling of import decls in overlays make sense?

Also update cursor info to report the owning module of the symbol under the cursor as the underlying/primary module when the symbol comes from a cross-import overlay.

Resolves rdar://problem/59445385

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci test Linux

**Comment by CodaFi** [comment]
@swift-ci test Windows

File: test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.response [hunk 1]
```diff
@@ -0,0 +1,313 @@
+import SwiftOnoneSupport
+
+public func fromA()
+
+
+// MARK: - B Additions
+
+import B
+
+// Available when B is imported with A
```
Comment (akyrtzi) [inline test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.response]: How does this synthesized comment relate to a user documentation comment, could you add some doc-comments in the test cases?

File: lib/IDE/ModuleInterfacePrinting.cpp [hunk 2]
```diff
@@ -238,6 +238,295 @@ swift::ide::findGroupNameForUSR(ModuleDecl *M, StringRef USR) {
   return None;
 }
 
+/// Prints a single decl using the \p Printer and \p Options provided.
+///
+/// \returns Whether the given decl was printed.
+static bool printModuleInterfaceDecl(Decl *D,
```
Comment (nathawes) [inline lib/IDE/ModuleInterfacePrinting.cpp]: This isn't new code - it's just pulled out of `printSubmoduleInterface()` below.

**Comment by nathawes** [comment]
@swift-ci please test

File: test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.response (hunk 1 above)
Comment (nathawes) [reply test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.response]: The synthesized comment is printed before it at the moment. Added a test case showing the current behavior.

File: test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.A.response [hunk 3]
```diff
@@ -0,0 +1,313 @@
+import SwiftOnoneSupport
+
+public func fromA()
+
+
+// MARK: - B Additions
+
+import B
+
+// Available when B is imported with A
```
Comment (akyrtzi) [inline test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.A.response]: Doesn't this seem a bit annoying to repeat it for every declaration? What do you think about only mention this once at the `MARK:` line?

Comment (nathawes) [reply test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.A.response]: @brentdax asked for it on every declaration in the radar so I'll redirect the question there. Brent, are you okay with Argyrios' suggestion?
It looks repetitive to me, but I'm also not sure how likely users are to scroll up in larger overlays if they get to one of these symbols by jump-to-definition or cmd+f.

Comment (akyrtzi) [reply test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.A.response]: > but I'm also not sure how likely users are to scroll up in larger overlays if they get to one of these symbols by jump-to-definition or cmd+f.
This is a good point, I retract my suggestion.

**Review by akyrtzi — APPROVED** [review]

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@brentdax I'm going to go ahead and merge this so I can land a few subsequent PRs based on some the APIs added in this one, but it'd be great if you could take a look when you get the chance (I know you're probably busy build czar-ing). I'm not sure I've used the right terminology everywhere, and am probably missing some interesting test cases.

**Pull request merged**